File: Week_01/G20190343020228/641.设计循环双端队列.cs
/*
 * @lc app=leetcode.cn id=641 lang=csharp
 *
 * [641] 设计循环双端队列
 */

// @lc code=start
public class MyCircularDeque {

    private int[] myqueue;
    private int head;
    private int size;
    /** Initialize your data structure here. Set the size of the deque to be k. */
    public MyCircularDeque(int k) {
        myqueue = new int[k];
        head=0;
        size=0;
    }
    
    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    public bool InsertFront(int value) {
        if(IsFull()) return false;
        head = (head+myqueue.Length-1)%myqueue.Length;
        myqueue[head]=value;
        size++;
        return true;
    }
    
    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    public bool InsertLast(int value) {
        if(IsFull()) return false;
        myqueue[(head+size)%myqueue.Length]=value;
        size++;
        return true;
    }
    
    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    public bool DeleteFront() {
        if(IsEmpty()) return false;
        head = (head+1)%myqueue.Length;
        size--;
        return true;
    }
    
    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    public bool DeleteLast() {
        if(IsEmpty()) return false;
        size--;
        return true;
    }
    
    /** Get the front item from the deque. */
    public int GetFront() {
        return size!=0?myqueue[head]:-1;
    }
    
    /** Get the last item from the deque. */
    public int GetRear() {
        return size != 0 ? myqueue[(head + size - 1) % myqueue.Length] : -1;
    }
    
    /** Checks whether the circular deque is empty or not. */
    public bool IsEmpty() {
        return size==0;
    }
    
    /** Checks whether the circular deque is full or not. */
    public bool IsFull() {
        return size==myqueue.Length;
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque obj = new MyCircularDeque(k);
 * bool param_1 = obj.InsertFront(value);
 * bool param_2 = obj.InsertLast(value);
 * bool param_3 = obj.DeleteFront();
 * bool param_4 = obj.DeleteLast();
 * int param_5 = obj.GetFront();
 * int param_6 = obj.GetRear();
 * bool param_7 = obj.IsEmpty();
 * bool param_8 = obj.IsFull();
 */
// @lc code=end

File: Week_01/G20190343020102/641.设计循环双端队列.js
/*
 * @lc app=leetcode.cn id=641 lang=javascript
 *
 * [641] 设计循环双端队列
 */

// @lc code=start
/**
 * Initialize your data structure here. Set the size of the deque to be k.
 * @param {number} k
 */


 // 这题不会，剽窃的代码
var MyCircularDeque = function (k) {
  this.ArrDeque = new Array(k + 1);
  this.front = 0;
  this.rear = 0;
  this.n = k + 1;
  this.k = k;
};


/**
 * Adds an item at the front of Deque. Return true if the operation is successful. 
 * @param {number} value
 * @return {boolean}
 */
MyCircularDeque.prototype.insertFront = function (value) {
  if (this.isFull('front')) {
    return false;
  } else {
    this.front = (--this.front + this.n) % (this.n);
    this.ArrDeque[this.front] = value;
    return true;
  }
}

/**
 * Adds an item at the rear of Deque. Return true if the operation is successful. 
 * @param {number} value
 * @return {boolean}
 */
MyCircularDeque.prototype.insertLast = function (value) {
  if (this.isFull('last')) {
    return false;
  } else {
    this.ArrDeque[this.rear] = value;
    this.rear = (this.rear + 1) % (this.n);
    return true;
  }
};

/**
 * Deletes an item from the front of Deque. Return true if the operation is successful.
 * @return {boolean}
 */
MyCircularDeque.prototype.deleteFront = function () {
  if (this.isEmpty()) {
    return false;
  } else {
    this.front = (++this.front) % (this.n);
    return true;
  }
};

/**
 * Deletes an item from the rear of Deque. Return true if the operation is successful.
 * @return {boolean}
 */
MyCircularDeque.prototype.deleteLast = function () {
  if (this.isEmpty()) {
    return false
  } else {
    this.rear = (--this.rear + this.n) % (this.n);
    return true;
  }

};

/**
 * Get the front item from the deque.
 * @return {number}
 */
MyCircularDeque.prototype.getFront = function () {
  return this.isEmpty() ? -1 : this.ArrDeque[this.front]
};

/**
 * Get the last item from the deque.
 * @return {number}
 */
MyCircularDeque.prototype.getRear = function () {
  return this.isEmpty() ? -1 : this.ArrDeque[(this.rear - 1 + this.n) % (this.n)]
};

/**
 * Checks whether the circular deque is empty or not.
 * @return {boolean}
 */
MyCircularDeque.prototype.isEmpty = function () {
  if (this.front == this.rear) {
    return true;
  } else {
    return false;
  }
};

/**
 * Checks whether the circular deque is full or not.
 * @return {boolean}
 */
MyCircularDeque.prototype.isFull = function () {
  return (this.rear - this.front + this.n) % (this.n) == this.k;
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * var obj = new MyCircularDeque(k)
 * var param_1 = obj.insertFront(value)
 * var param_2 = obj.insertLast(value)
 * var param_3 = obj.deleteFront()
 * var param_4 = obj.deleteLast()
 * var param_5 = obj.getFront()
 * var param_6 = obj.getRear()
 * var param_7 = obj.isEmpty()
 * var param_8 = obj.isFull()
 */
// @lc code=end

